File: internal/obs/heartbeat.go
package obs

import (
	"os"
	"strconv"
	"time"
)

// EnvHeartbeatIntervalMs is the environment variable the lord injects into a thrall so its SDK
// heartbeats at the manifest-configured interval; the lord derives its reaper threshold from the
// same value, so the two never drift.
const EnvHeartbeatIntervalMs = "AETHER_HEARTBEAT_INTERVAL_MS"

const (
	// DefaultHeartbeatIntervalMs is the heartbeat interval when unset (the historical 2s).
	DefaultHeartbeatIntervalMs = 2000
	// MinHeartbeatIntervalMs floors the interval so a typo cannot flood the bus / peg the CPU.
	MinHeartbeatIntervalMs = 100
)

// ClampHeartbeatIntervalMs normalizes a configured interval in milliseconds: a non-positive
// value falls back to the default, and a too-small value is raised to the minimum floor.
func ClampHeartbeatIntervalMs(ms int) int {
	if ms <= 0 {
		return DefaultHeartbeatIntervalMs
	}
	if ms < MinHeartbeatIntervalMs {
		return MinHeartbeatIntervalMs
	}
	return ms
}

// HeartbeatInterval resolves the SDK heartbeat interval from the environment (empty or invalid
// -> default, clamped to the minimum). Used by the Go SDK; the TS/Python SDKs mirror it.
func HeartbeatInterval() time.Duration {
	ms, err := strconv.Atoi(os.Getenv(EnvHeartbeatIntervalMs))
	if err != nil {
		ms = 0
	}
	return time.Duration(ClampHeartbeatIntervalMs(ms)) * time.Millisecond
}

File: sdk/ts/src/heartbeat.ts
// Heartbeat interval resolution, mirroring internal/obs: the lord injects
// AETHER_HEARTBEAT_INTERVAL_MS and the SDK beats at it, so the lord's reaper threshold (derived
// from the same value) and the thralls never drift. An empty/invalid value falls back to the
// default; a too-small value is clamped to a floor.

export const DEFAULT_HEARTBEAT_INTERVAL_MS = 2000;
export const MIN_HEARTBEAT_INTERVAL_MS = 100;

export function clampHeartbeatIntervalMs(ms: number): number {
  if (!Number.isFinite(ms) || ms <= 0) return DEFAULT_HEARTBEAT_INTERVAL_MS;
  if (ms < MIN_HEARTBEAT_INTERVAL_MS) return MIN_HEARTBEAT_INTERVAL_MS;
  return ms;
}

export function heartbeatIntervalMs(): number {
  const raw = process.env.AETHER_HEARTBEAT_INTERVAL_MS;
  return clampHeartbeatIntervalMs(raw ? parseInt(raw, 10) : NaN);
}
